halo2/src/circuit_builder.rs

```diff
@@ -316,17 +379,19 @@ impl<'a, T: FieldElement, F: PrimeField<Repr = [u8; 32]>> Circuit<F> for PowdrCi
 
                 // Set witness values
                 let mut public_cells = Vec::new();
+                let witness: Option<&[(String, Vec<T>)]> = if !new_witness.is_empty() {
+                    Some(&new_witness)
+                } else {
+                    self.witness
+                };
                 for (name, &column) in config.advice.iter() {
                     // Note that we can't skip this loop if we don't have a witness,
                     // because the copy_advice() call below also adds copy constraints,
                     // which are needed to compute the verification key.
-                    let values = self.witness.as_ref().map(|witness| {
-                        witness
-                            .iter()
-                            .find_map(|(witness_name, values)| {
-                                (witness_name == name).then_some(values)
-                            })
-                            .unwrap_or_else(|| panic!("Missing witness: {}", name))
```

Comment:
Now, there might be columns missing if we're not in the latest phase. That is fine though, Halo2 complaints if there are still unassigned witnesses by the last phase.